libtenzir/builtins/functions/ceil_floor.cpp

```diff
@@ -0,0 +1,168 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/arrow_time_utils.hpp>
+#include <tenzir/arrow_utils.hpp>
+#include <tenzir/concept/parseable/tenzir/si.hpp>
+#include <tenzir/detail/narrow.hpp>
+#include <tenzir/series_builder.hpp>
+#include <tenzir/tql2/eval.hpp>
+#include <tenzir/tql2/plugin.hpp>
+
+#include <arrow/util/tdigest.h>
+
+namespace tenzir::plugins::numeric {
+
+TENZIR_ENUM(mode, ceil, floor, round);
+
+namespace {
+template <mode Mode>
+class plugin final : public function_plugin {
+public:
+  auto name() const -> std::string override {
+    return std::string{to_string(Mode)};
+  }
+
+  auto make_function(invocation inv,
+                     session ctx) const -> failure_or<function_ptr> override {
+    auto expr = ast::expression{};
+    auto spec = std::optional<located<duration>>{};
+    TRY(argument_parser2::function(function_name())
+          .add(expr, "<value>")
+          .add(spec, "<spec>")
+          .parse(inv, ctx));
+    if (spec && spec->inner.count() == 0) {
+      diagnostic::error("resolution must not be 0")
+        .primary(spec.value())
+        .emit(ctx);
+      return failure::promise();
+    }
+    return function_use::make([expr = std::move(expr), spec = std::move(spec),
+                               this](evaluator eval, session ctx) -> series {
+      const auto value = located{eval(expr), expr.get_location()};
+      const auto& ty = value.inner.type;
+      const auto length = value.inner.length();
+      if (not spec) {
+        // fn(<number>)
+        const auto f = detail::overload{
+          [&](const arrow::NullArray&) {
+            return series::null(ty, length);
+          },
+          [&]<concepts::one_of<arrow::Int64Array, arrow::UInt64Array> T>(
+            const T&) {
+            return value.inner;
+          },
+          [&](const arrow::DoubleArray& arg) {
+            auto b = arrow::DoubleBuilder{};
+            check(b.Reserve(length));
+            for (auto row = int64_t{0}; row < length; ++row) {
+              if (arg.IsNull(row) or not std::isfinite(arg.Value(row))) {
+                check(b.AppendNull());
+                continue;
+              }
+              if constexpr (Mode == mode::ceil) {
+                check(b.Append(std::ceil(arg.Value(row))));
+              } else if constexpr (Mode == mode::floor) {
+                check(b.Append(std::floor(arg.Value(row))));
+              } else {
+                check(b.Append(std::round(arg.Value(row))));
+              }
+            }
+            return series{ty, finish(b)};
```

Comment:
These functions do not convert double to Int64 as it may not be within the range of the type. Do we want to keep the original behavior?

Reply:
We discussed this in a huddle. From an information theoretical standpoint, keeping a double is better, but in 99% of the cases an integer is what the user wants. Thus we'll go with `int64` as the return type here when the input is `double`.